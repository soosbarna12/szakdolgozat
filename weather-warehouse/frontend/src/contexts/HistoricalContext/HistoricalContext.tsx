import React, { createContext, useEffect, useState } from "react";
import { LOCAL_STORAGE_TEMPERATURE_SCALE } from "../../consts/temperatureScale.const";
import { TemperatureScale } from "../../types/temperatureScale.type";
import { HistoricalContextType, HistoricalLocationData, HistoricalPageData, HistoricalProviderProps } from "./HistoricalContext.type";


export const HistoricalContext = createContext<HistoricalContextType>({
  location: { name: "", lat: 0, lon: 0 },
  setLocation: () => { },
  historicalPageData: [] as any,
  setHistoricalPageData: () => { },
  temperatureScale: TemperatureScale.Celsius,
  setTemperatureScale: () => { },
});

export const HistoricalLocationProvider = ({ children }: HistoricalProviderProps) => {

  const [location, setLocation] = useState<HistoricalLocationData>(() => {
    const storedLocation = localStorage.getItem("historicalLocation");
    const defaultLocation = { name: "", lat: 0, lon: 0 };
    try {
      return storedLocation ? JSON.parse(storedLocation) : defaultLocation;
    } catch (error) {
      console.error("Failed to parse location from localStorage:", error);
      return defaultLocation;
    }
  });

  const [historicalPageData, setHistoricalPageData] = useState<HistoricalPageData[]>(() => {
    const storedData = localStorage.getItem("historicalPageData");
    try {
      return storedData ? JSON.parse(storedData) : [];
    } catch (error) {
      console.error("Failed to parse historical data from localStorage:", error);
      return [];
    }
  });

  const [temperatureScale, setTemperatureScale] = useState<TemperatureScale | string>(() => {
    const storedTemperatureScale = localStorage.getItem(LOCAL_STORAGE_TEMPERATURE_SCALE);
    try {
      return storedTemperatureScale ? JSON.parse(storedTemperatureScale) : TemperatureScale.Celsius;
    } catch (error) {
      console.error("Failed to parse temperature scale from localStorage:", error);
      return TemperatureScale.Celsius;
    }
  });


  useEffect(() => {
    if (historicalPageData) {
      localStorage.setItem("historicalPageData", JSON.stringify(historicalPageData));
    }
  }, [historicalPageData]);

  useEffect(() => {
    localStorage.setItem("historicalLocation", JSON.stringify(location));
  }, [location]);

  useEffect(() => {
    localStorage.setItem(LOCAL_STORAGE_TEMPERATURE_SCALE, JSON.stringify(temperatureScale));
  }, [temperatureScale]);


  return (
    <HistoricalContext.Provider value={{ location, setLocation, historicalPageData, setHistoricalPageData, temperatureScale, setTemperatureScale }}>
      {children}
    </HistoricalContext.Provider>
  );
};
