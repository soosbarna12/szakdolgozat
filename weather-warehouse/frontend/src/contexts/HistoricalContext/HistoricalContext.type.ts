import { Dispatch, ReactNode, SetStateAction } from "react";
import { HistoricalDataTable } from "../../types/historicalDataTable.type";
import { TemperatureScale } from "../../types/temperatureScale.type";

// location name interfere
export interface HistoricalLocationData {
  name: string;
  country?: string;
  lat: number;
  lon: number;
}

// historical page data
// eslint-disable-next-line @typescript-eslint/no-empty-object-type
export interface HistoricalPageData extends HistoricalDataTable {}

export interface HistoricalContextType {
  location: HistoricalLocationData;
  historicalPageData: any []; // add historicalData to the context type
  temperatureScale: TemperatureScale | string;
  setLocation: Dispatch<SetStateAction<HistoricalLocationData>>;
  setHistoricalPageData: Dispatch<SetStateAction<HistoricalPageData[]>>; // add setter for historicalData
  setTemperatureScale: Dispatch<SetStateAction<TemperatureScale | string>>;
}

export interface HistoricalProviderProps {
  children: ReactNode;
}
