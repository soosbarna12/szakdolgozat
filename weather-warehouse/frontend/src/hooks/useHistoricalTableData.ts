import { useContext, useEffect } from "react";
import { HistoricalContext } from "../contexts/HistoricalContext/HistoricalContext";
import { UseHistoricalDataProps } from "../pages/HistoricalPage/HistoricalPage.type";


export function useHistoricalTableData(props: Readonly<UseHistoricalDataProps>) {
  //const [tableData, setTableData] = useState<HistoricalDataTable[]>([]);
  const { historicalPageData, setHistoricalPageData } = useContext(HistoricalContext);
  const { data: newRecord } = props;

  useEffect(() => {
    if (newRecord) {
      setHistoricalPageData((prev) => {
        const isDuplicate = prev.some((record) => record.date === newRecord.date
        && record.cityName === newRecord.cityName
        && record.countryCode === newRecord.countryCode);
        return isDuplicate ? prev : [...prev, newRecord];
      });
    }
  }, [newRecord]);

  return { historicalPageData, setHistoricalPageData };
}
