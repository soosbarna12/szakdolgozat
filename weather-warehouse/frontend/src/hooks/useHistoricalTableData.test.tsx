import '@testing-library/jest-dom';
import { act, renderHook } from '@testing-library/react';
import React from 'react';
import { HistoricalContext } from '../contexts/HistoricalContext/HistoricalContext';
import { useHistoricalTableData } from './useHistoricalTableData';

describe('useHistoricalTableData', () => {
  const mockSetHistoricalPageData = jest.fn();
  const mockHistoricalPageData = [
    { date: '2023-01-01', cityName: 'New York', countryCode: 'US' },
    { date: '2023-01-02', cityName: 'Los Angeles', countryCode: 'US' },
  ];

  const wrapper = ({ children }: { children: React.ReactNode }) => (
    <HistoricalContext.Provider value={{
      historicalPageData: mockHistoricalPageData,
      setHistoricalPageData: mockSetHistoricalPageData,
    } as any}
    >
      {children}
    </HistoricalContext.Provider>
  );

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches useHistoricalTableData hook snapshot', () => {
    const { result } = renderHook(() =>
      useHistoricalTableData({ data: null } as any),
      { wrapper }
    );

    expect(result.current).toMatchSnapshot();
  });

  it('adds a new record to historicalPageData if it is not a duplicate', () => {
    const newRecord = {
      date: '2023-01-03',
      cityName: 'Chicago',
      countryCode: 'US',
    };

    renderHook(() => useHistoricalTableData({ data: newRecord } as any), { wrapper });

    expect(mockSetHistoricalPageData.mock.calls[0][0](mockHistoricalPageData)).toEqual([
      ...mockHistoricalPageData,
      newRecord,
    ]);
  });

  it('does not add a duplicate record to historicalPageData', () => {
    const duplicateRecord = {
      date: '2023-01-01',
      cityName: 'New York',
      countryCode: 'US',
    };

    renderHook(() => useHistoricalTableData({ data: duplicateRecord } as any), {
      wrapper,
    });

    act(() => {
      mockSetHistoricalPageData.mock.calls[0][0](mockHistoricalPageData);
    });

    expect(mockSetHistoricalPageData).not.toHaveBeenCalledWith([
      ...mockHistoricalPageData,
      duplicateRecord,
    ]);
  });

  it('does nothing if newRecord is null', () => {
    renderHook(() => useHistoricalTableData({ data: null } as any), { wrapper });

    expect(mockSetHistoricalPageData).not.toHaveBeenCalled();
  });
});