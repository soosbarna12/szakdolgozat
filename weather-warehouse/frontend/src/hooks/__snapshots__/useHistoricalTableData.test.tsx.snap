// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useHistoricalTableData matches useHistoricalTableData hook snapshot 1`] = `
{
  "historicalPageData": [
    {
      "cityName": "New York",
      "countryCode": "US",
      "date": "2023-01-01",
    },
    {
      "cityName": "Los Angeles",
      "countryCode": "US",
      "date": "2023-01-02",
    },
  ],
  "setHistoricalPageData": [MockFunction],
}
`;
